src/geometry/euclidean/ceva_theorem.lean

```diff
@@ -0,0 +1,325 @@
+  have hs := S.independent,
+  set T : affine_basis (fin 3) ℝ V := ⟨S.points, S.independent, hspan⟩ with hT,
+  replace h₈ : O ∈ interior (convex_hull ℝ (set.range T.points)) := by exact h₈,
+  set σ₁ : perm (fin 3) := equiv.refl (fin 3) with hσ₁,
+  set σ₂ : perm (fin 3) := list.form_perm [0, 1, 2] with hσ₂,
+  set σ₃ : perm (fin 3) := equiv.trans σ₂ σ₂ with hσ₃,
+  set S₁ : triangle ℝ V := ⟨![A, B, C] ∘ σ₁, by simpa [affine_independent_equiv, ← h₁]⟩ with hS₁,
+  have hS₁span := affine_span.simplex_eq_top S₁ h₀,
+  set T₁ : affine_basis (fin 3) ℝ V := ⟨S₁.points, S₁.independent, hS₁span⟩ with hT₁,
+  have hTσ₁ : T₁.points = T.points ∘ σ₁ := by simp [h₁],
+  replace h₂ : collinear ℝ ({S₁.points 0, S₁.points 1, D} : set V) := by convert h₂,
+  replace h₅ : collinear ℝ ({D, O, S₁.points 2} : set V) := by convert h₅,
+  set S₂ : triangle ℝ V := ⟨![A, B, C] ∘ σ₂, by simpa [affine_independent_equiv, ← h₁]⟩ with hS₂,
+  have hS₂span := affine_span.simplex_eq_top S₂ h₀,
+  set T₂ : affine_basis (fin 3) ℝ V := ⟨S₂.points, S₂.independent, hS₂span⟩ with hT₂,
+  have hTσ₂ : T₂.points = T.points ∘ σ₂ := by simp [h₁],
+  replace h₃ : collinear ℝ ({S₂.points 0, S₂.points 1, E} : set V) := by convert h₃,
+  replace h₆ : collinear ℝ ({E, O, S₂.points 2} : set V) := by convert h₆,
+  set S₃ : triangle ℝ V := ⟨![A, B, C] ∘ σ₃, by simp only [affine_independent_equiv, ← h₁,
+    S.independent]⟩ with hS₃,
+  have hS₃span := affine_span.simplex_eq_top S₃ h₀,
+  set T₃ : affine_basis (fin 3) ℝ V := ⟨S₃.points, S₃.independent, hS₃span⟩ with hT₃,
+  have hTσ₃ : T₃.points = T.points ∘ σ₃ := by simp [h₁],
+  replace h₄ : collinear ℝ ({S₃.points 0, S₃.points 1, F} : set V) := by convert h₄,
+  replace h₇ : collinear ℝ ({F, O, S₃.points 2} : set V) := by convert h₇,
+  have hwnezero : T.coord 0 O * T.coord 1 O * T.coord 2 O ≠ 0,
+  { apply ne_of_gt,
+    simp only [mul_pos, affine_basis.interior_coord_pos h₈] },
+  have hADB := affine_basis.fin3_interior_coord_mul_dist_eq σ₁ T T₁ hTσ₁
+    ( affine_basis_fin3_coord_vsub_smul_sum_eq_zero O D T₁ h₂ h₅) h₈,
+  have hBEC := affine_basis.fin3_interior_coord_mul_dist_eq σ₂ T T₂ hTσ₂
+    (affine_basis_fin3_coord_vsub_smul_sum_eq_zero O E T₂ h₃ h₆) h₈,
+  have hCFA := affine_basis.fin3_interior_coord_mul_dist_eq σ₃ T T₃ hTσ₃
+    (affine_basis_fin3_coord_vsub_smul_sum_eq_zero O F T₃ h₄ h₇) h₈,
+  clear h₂ h₃ h₄ h₅ h₆ h₇ h₈,
+  have hs₁ : σ₂ 0 = 1 := by refl,
+  have hs₂ : σ₂ 1 = 2 := by refl,
+  have hs₃ : σ₂ 2 = 3 := by refl,
+  have hb : ![A, B, C] 1 = B := by refl,
+  have hc : ![A, B, C] 2 = C := by refl,
+  have ha : ![A, B, C] 3 = A := by refl,
+  have h := congr_arg2 (λ a b, a * b) (congr_arg2 (λ a b, a * b) hADB hBEC) hCFA,
+  simp only [← affine_basis.coord_perm σ₁ T T₁ O hTσ₁, ← affine_basis.coord_perm σ₂ T T₂ O hTσ₂,
+    ← affine_basis.coord_perm σ₃ T T₃ O hTσ₃] at h,
+  clear hADB hBEC hCFA hTσ₁ hTσ₂ hTσ₃ hT₁ hT₂ hT₃ T₁ T₂ T₃ hS₁span hS₂span hS₃span,
+  dsimp at h, -- non-terminal dsimp, do not know how to fix it
```

Comment:
Non-terminal `dsimp`s are not so bad. Because the `term_before` and the `term_after` should still be defeq. Which means most tactics don't care too much about what the `dsimp` did. Also, there are far fewer lemmas that `dsimp` uses, and there are often very straightforward. Small chance that a non-terminal `dsimp` will break a proof in a bad way in the future.